### PR TITLE
WWST-1595 Zigbee Accessory Dimmer

### PR DESCRIPTION
**Pull request by greens**
Primary intention is to work with the Aurora Wireless Wall Remote

### CODE REVIEW

**Comment by greens** [comment]
Here's the first draft.

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy]: aborted attempts at the faked intermediate step events

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy:85]: we're essentially turning the middle buttons (when held) into "go to max" and "go to min"

Comment (greens) [reply devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy:85]: with nothing in-between

fine tuning can be handled by the single press events

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy:64]: This device has explicit On and Off buttons, correct? If so I think it may be better to create the `on` event on receipt of the `On` command rather than toggling. Should also force each event to be a state change event.

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy]: We usually don't create level events for a value of 0 because when they are turned on it is still 0.

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy]: Should add `isStateChange: true` to all these events

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy]: I don't think these configs will do anything since the device doesn't have On/Off or Level cluster servers, just clients.

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy:24]: If we want to add Health Check might be able to setup attribute reporting on any implemented attribute in the Basic cluster

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (greens) [reply devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy:64]: Only problem is that then we have to maintain the state and there's no way to get that from the device. (and it's not indicated on the device anywhere)

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (greens) [reply devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy]: I did a little futzing about to make sure that if you turn it on from level = 0, level goes up to 10, but avoiding that situation in the first place is probably the better choice.

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (greens) [reply devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy:24]: I tried doing read attributes of basic cluster, but wasn't getting a response.

**Comment by greens** [comment]
Updated. Still can't get the device to report any attributes in the basic cluster, so I don't think this device can support health check.

**Comment by greens** [comment]
Waiting for a final review from somebody.

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy]: Can remove `minHubCoreVersion` and `executeCommandsLocally` since it doesn't run locally.

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (tpmanley) [reply devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy:64]: Does this device have separate on and off buttons or is it just a single button that toggles between on/off?

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy]: Need to set `switch` to `off` if level is 0

**Comment by greens** [comment]
Done.

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (greens) [reply devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy:64]: a single button

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy]: I think this `else` can be removed

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy]: Could simplify to just `else`

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy]: We usually don't set level to 0

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy]: Usually the default level is 100

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy:60]: Does it use both of these commands? Seems odd if it has one button that can generate two different commands.

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy]: Guessing you could call `zigbee. addBinding(0x0006)`, etc.... Since this device doesn't have reportable attributes it's only the binding that should matter.

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (greens) [reply devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy:60]: It does, but I think it's for talking to actual devices that are keeping on/off so it keeps track of state internally

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (greens) [reply devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy]: yep, that was it

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (varzac) [inline devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy]: You can remove this commented line.  You could also change the above to use `zigbee.ONOFF_CLUSTER`, `zigbee.LEVEL_CONTROL_CLUSTER` for the first 2.  We don't have a constant currently for the scenes cluster though.

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (varzac) [reply devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy:60]: I'm wondering if we want to track state separately or just use what the device gives us.  If they have other bulbs that they are controlling directly from the remote it would be weird for us to be generating an "off" event when the bulbs are actually getting an on.

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (varzac) [reply devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy:64]: As I commented above, we can use the command it sends us to know.  If the command ID is `0x01` it's an on and `0x00` is an off.  I think we could consider the argument that we should explicitly mark the event with `isStateChange: true` as well, but I don't feel too strongly on that.

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (tpmanley) [reply devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy:60]: I agree, if it's tracking state on it's own I think we should use that.

Comment (greens) [reply devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy:60]: The intended operation with Aurora's hub is to control other lights, I assume.

I don't know any way they'd be able to do this on our system, though. And there's no outward indication on the device or any way for us to determine it's internal state.

Tracking the device's on/off state, but just having a purely invented level value seems counterintuitive to me and also removes the ability to use the app as a "virtual" on/off switch.

Comment (tpmanley) [reply devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy:60]: 👍  I see your point

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy:125]: I think you need to create a `numberOfButtons` event too.

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy]: nit: indentation looks off with the line below

File: devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy:117]: nit: indentation looks off

**Review by tpmanley — DISMISSED** [review]
Just a couple of final minor things and then it looks good

**Review by tpmanley — APPROVED** [review]

**Pull request merged**
